Add a new LyricsWindow events plugin

CreamyCookie opened the pull request:
This plugin (an EventPlugin) is based on the [LyricWiki plugin](https://github.com/quodlibet/quodlibet/blob/master/quodlibet/quodlibet/ext/songsmenu/lyricwiki.py) (a SongsMenuPlugin), which it aims to replace.

Compared to it, it has the following advantages:

* The window gets opened when Quod Libet is launched. Closes  #1629.
* The user can configure the zoom level of the web view.
* Window position and size are (re)stored.

Additionally I might implement #1485 and parts or the entirety of #1628 in a later pull request, which would make the plugin use a search engine when there are no lyrics in LyricWikia. This is also the reason why it was renamed to **lyricswindow.py**. 

This is how it looks with my current configuration (zoom = 1.4):

![lyricswindow](https://cloud.githubusercontent.com/assets/3063858/10487469/636a3a28-7294-11e5-8089-d8cd2833ff86.png)

---

One thing I'd like to do, but don't really know how, is to set `LyricsWebViewWindow.set_transient_for(quod_libet_main_window)` ([doc](http://www.pygtk.org/pygtk2reference/class-gtkwindow.html#method-gtkwindow--set-transient-for)), so that when bringing Quod Libet to the front, the same would happen to the lyrics window. How do I get main window in the plugin?

Regarding the Copyright notice: Can I leave it the way it is?

lazka commented:
> This plugin (an EventPlugin) is based on the LyricWiki plugin (a SongsMenuPlugin), which it aims to replace.

I guess we should allow merging EventPlugins and SongsMenuPlugin so we can have both enabled() and the song menu actions.

> Compared to it, it has the following advantages:
>
>    The window gets opened when Quod Libet is launched. Closes #1629.

OK

>    The user can configure the zoom level of the web view.

Why not hardcode a good value if it is too small?

>    Window position and size are (re)stored.

There is qltk.window.PersistentWindowMixin for that as well. But it's hardcoded to same to "memory" atm, and we don't want plugin state there.

> Additionally I might implement #1485 and parts or the entirety of #1628 in a later pull request, which would make the plugin use a search engine when there are no lyrics in LyricWikia. This is also the reason why it was renamed to lyricswindow.py.

Sounds good.

> This is how it looks with my current configuration (zoom = 1.4):

> One thing I'd like to do, but don't really know how, is to set LyricsWebViewWindow.set_transient_for(quod_libet_main_window) (doc), so that when bringing Quod Libet to the front, the same would happen to the lyrics window. How do I get main window in the plugin?

```
from quodlibet import app
app.window
```
SongMenuPlugins have a ``plugin_window`` attribute set to the window which owns the menu.

> Regarding the Copyright notice: Can I leave it the way it is?

You should add your name

lazka commented:
> One thing I'd like to do, but don't really know how, is to set LyricsWebViewWindow.set_transient_for(quod_libet_main_window) (doc), so that when bringing Quod Libet to the front, the same would happen to the lyrics window. How do I get main window in the plugin?

Btw. you linked to gtk+2 docs. Here are the gtk+3 ones: https://lazka.github.io/pgi-docs/#Gtk-3.0/classes/Window.html#Gtk.Window.set_transient_for

CreamyCookie commented:
Okay, I updated the code according to your feedback. :)

> I guess we should allow merging EventPlugins and SongsMenuPlugin so we can have both enabled() and the song menu actions.

Yeah, that would be useful. In regards to this I changed the `plugin_on_song_started` event, so it dosen'- reopen the window if it's closed, when a new song starts.

> There is qltk.window.PersistentWindowMixin for that as well. But it's hardcoded to same to "memory" atm, and we don't want plugin state there.

What would need to change in `PersistentWindowMixin` to make it usable in this case?

> Btw. you linked to gtk+2 docs. Here are the gtk+3 ones: https://lazka.github.io/pgi-docs/#Gtk-3.0/classes/Window.html#Gtk.Window.set_transient_for

Ah nice. Thank you! :+1:

lazka commented:
OK, thanks and sorry for the delay. There are still a few no-gos. The Window can't be reopened easily and having two plugins doing the same thing but in slightly different way is confusins (plus the code duplication).

I'll look into fixing PersistentWindowMixin and to allow making a mixed songmenu/event plugins before continuing here.

CreamyCookie commented:
@lazka Okay, cool. I will delete the original plugin file in my branch. (I first wanted to get feedback before doing that.)

I might add the search function, which I mentioned earlier, in the meantime.

> The Window can't be reopened easily

I guess this would be possible when PersistentWindowMixin is fixed, right?

CreamyCookie commented:
@lazka I have added the search functionality and refactored the code a bit. (lyricswindow and another plugin were using the same classes.) That took way too much time, haha.. But hey, I learned the basics of git.

Please feel free to code review the heck out of this. :smile:

declension commented on quodlibet/quodlibet/plugins/__init__.py:
```diff
@@ -500,3 +500,69 @@ def ConfigCheckButton(cls, label, name, default=False):
         return ConfigCheckButton(label, PM.CONFIG_SECTION,
                                  option, populate=True)
+
+
```
Unless I've missed the point, this seems like a lot of OOP to do what `PluginConfigMixin` is already doing (`config_get_bool`, `config_set` etc)?

declension commented on quodlibet/quodlibet/ext/events/lyricswindow.py:
```diff
@@ -0,0 +1,368 @@
+
+    @cached_property
+    def Conf(self):
+        return get_config(self.PLUGIN_ID)
```
This now seems very similar to what `PluginConfigMixin` does, essentially (but less configurable for the config section actually)...

CreamyCookie commented:
Hmm, would be cool, if GitHub allowed me to directly respond on a certain "line note".

> Unless I've missed the point, this seems like a lot of OOP to do what PluginConfigMixin is already doing (config_get_bool, config_set etc)?

I guess that's true. There are a few advantages to the `ConfProp` classes though:

* You can just access and set them like normal attributes (e.g. `conf.zoom_level = 3`).
* You don't need to know the exact type. (You don't need to care whether `zoom_level` is a float or int.) 
* The code seems more readable, because it isn't littered with `config_get`'s.
* It allows your IDE to provide code completion.


Note that these classes weren't my idea. They came from the [Animated On-Screen Display plugin](https://github.com/quodlibet/quodlibet/blob/master/quodlibet/quodlibet/ext/events/animosd/config.py). Hmm.. I think there's another advantage shown here, regarding `ColorConfProp`: You can easily extend the configuration classes.

I don't know if that's worth it though. So, if you give me the sign, I'll use `config_get_bool` etc. everywhere. I'd also add `config_get_float` then, if that's okay.

> This now seems very similar to what PluginConfigMixin does, essentially (but less configurable for the config section actually)...

What do you mean about being less configurable?

declension commented:
Hi. Yes it's annoying about line comment threads.

Anyway I see the benefits of the more OOP approach, to an extent (and looked again at Animated OSD, which is probably due a bit of refactoring anyway)... but with over 70 plugins in the codebase managing the diversity amongst them is quite a task sometimes, so I'm keen to get _more_ consistency first, and in the interests of getting the PR merged, keep _that_ work separate from extended the common config code significantly. So for now I'd be happiest if this used the `config_get_bool` etc (yes though please add related `float` / `int` ones if they're necessary)

Afterwards, I'm happy to revisit the, and fix, de-duplicate and improve the config system _across_ plugins as a separate commit / PR / Github Issue.

> Could you explain what you mean by the solution being less configurable?

Oh... it isn't that important, but it's down to the fallback / legacy support mechanism used (via `CONFIG_SECTION` in that class)

@lazka do you have any further comments?

lazka commented:
> @lazka do you have any further comments?

 re config? I don't  care much about how it's done. The animosd way is using descriptors because it used to use a special config object before and at the last refactor I just kept the interface as is and connected it to the config.

I've recently moved some plugins to PluginConfig (https://github.com/quodlibet/quodlibet/blob/master/quodlibet/quodlibet/ext/events/trayicon/util.py#L18) which acts as a proxy to the real config. I prefer it over the mixin but ymmv.

declension commented:
@lazka - meant about the PR entirely really (LGTM other than above)

Re: config - yes, I see now, that's quite cool too. I guess _any_ of these (three?) existing methods for config is probably good I guess, would be good to settle on one or two and migrate slowly, but outside of the scope for this here..

lazka commented:
I don't like losing the song menu item for this plugin. But I don't feel that strongly about it.. so feel free to merge.

Tamriel commented:
I would love to see this merged!
A lyrics window with custom zoom level is exactly what I wished for.

CreamyCookie commented:
> so feel free to merge.

Sorry for responding so late. How do I merge? I don't have write access.

---

For reference, the following are the tasks that should probably be done in other pull requests, because they would increase the complexity of this one even more.

* fix PersistentWindowMixin so that it can be used for plugins, instead of having to manually set the size / position of windows
* creating a class that inherits from both EventPlugins and SongsMenuPlugin, so the window can be easily reopened via the context menu of a song
* possibly decide for one method of storing configurations in plugins (ConfProp, PluginConfigMixin, and PluginConfig)

lazka commented:
> Sorry for responding so late. How do I merge? I don't have write access.

This was directed at @declension, only project members can merge.

> * fix PersistentWindowMixin so that it can be used for plugins, instead of having to manually (re)store the size / position of windows
* creating a class that inherits from both EventPlugins and SongsMenuPlugin, so the window can be easily (re)opened via the context menu of a song
* possibly decide for one method of storing configurations in plugins (ConfProp, PluginConfigMixin, and PluginConfig)
>
> Should I add these as bug reports?

Yes please, sounds good.

declension commented:
Yes - I'm largely happy (ongoing config decisions aside), but the branch is conflicting now, so it needs fixing first...

CreamyCookie commented:
@declension What needs fixing?

declension commented:
The branch has merge conflicts so it needs rebasing / merging with `master`, that's all

CreamyCookie commented:
Yay, I did it (my first git rebase as far as I remember). I squashed a few commits via `fixup` in the interactive rebase, so the commits are more clear.

CreamyCookie commented:
Yay, I did it (my first git rebase as far as I remember). I squashed a few commits via `fixup` in the interactive rebase, so the commits are more clear. It should be ready to merge into master now.

CreamyCookie commented:
Yay, I did it! :)

I squashed a few commits via `fixup` in the interactive rebase, so the commits are more clear.  Furthermore, the WebView shouldn't make any sounds on its own anymore. (There were some annoying audio ads that started autoplaying, which were driving me crazy, because I didn't know that they were coming from the lyrics window.)

@declension Is this alright?

declension commented:
Thanks
